server/tilesource/base.py

```diff
@@ -352,29 +576,217 @@ def getRegion(self, width=None, height=None, **kwargs):
-                if not isinstance(tileData, PIL.Image.Image):
-                    tileData = PIL.Image.open(BytesIO(tileData))
-                posX = x * metadata['tileWidth'] - left
-                posY = y * metadata['tileHeight'] - top
-                # Add each tile to the image.  PIL crops these if they are off
-                # the edge.
-                image.paste(tileData, (posX, posY),
-                            tileData if mode == 'RGBA' and
-                            tileData.mode == 'RGBA' else None)
-
+        for tile in self._tileIterator(iterInfo):
+            # Add each tile to the image.  PIL crops these if they are off the
+            # edge.
+            image.paste(tile['tile'], (tile['x'] - left, tile['y'] - top),
+                        tile['tile'] if mode == 'RGBA' and
+                        tile['tile'].mode == 'RGBA' else None)
         # Scale if we need to
+        width = int(math.floor(width))
+        height = int(math.floor(height))
         if width != regionWidth or height != regionHeight:
             image = image.resize(
                 (width, height),
                 PIL.Image.BICUBIC if width > regionWidth else
                 PIL.Image.LANCZOS)
         return self._encodeImage(image, **kwargs)
 
+    def getMagnification(self):
+        """
+        Get the magnification for the highest-resolution level.
+
+        :return: magnification, width of a pixel in mm, height of a pixel in mm.
+        """
+        return {
+            'magnification': None,
+            'mm_x': None,
+            'mm_y': None
+        }
+
+    def getMagnificationForLevel(self, level=None):
+        """
+        Get the magnification at a particular level.
+
+        :param level: None to use the maximum level, otherwise the level to get
+            the magification factor of.
+        :return: magnification, width of a pixel in mm, height of a pixel in mm.
+        """
+        mag = self.getMagnification()
+
+        if level is not None and self.levels and level != self.levels - 1:
+            mag['scale'] = 2.0 ** (self.levels - 1 - level)
+            if mag['magnification']:
+                mag['magnification'] /= mag['scale']
+            if mag['mm_x'] and mag['mm_y']:
+                mag['mm_x'] *= mag['scale']
+                mag['mm_y'] *= mag['scale']
+        if self.levels:
+            mag['level'] = level if level is not None else self.levels - 1
+        return mag
+
+    def getLevelForMagnification(self, magnification=None, exact=False,
+                                 mm_x=None, mm_y=None, rounding='round',
+                                 **kwargs):
+        """
+        Get the level for a specific magnifcation or pixel size.  If the
+        magnification is unknown or no level is sufficient resolution, and an
+        exact match is not requested, the highest level will be returned.
+          At least one of magnification, mm_x, and mm_y must be specified.  If
```

Comment:
Looks like you return the maximum resolution level if none of magnification, mm_x, or mm_y are provided. Update documentation accordingly,